### PR TITLE
Fixes #19615 - Hosts chart to c3

### PR DESCRIPTION
**Pull request by matanwerbner**
This PR simply changes the host configuration chart in the overview page from jquery flot to c3.
there is some refactoring done to chart related files, to be able to use them more generically.
this can be expanded to take out jquery.flot entirely in the future.

### CODE REVIEW

**Comment by ohadlevy** [comment]
@matanwerbner thanks, mind sending a before / after screenshot?

**Comment by matanwerbner** [comment]
Before:
screenshot 1
After:

**Comment by matanwerbner** [comment]
Before:
screenshot 1
After:
screenshot

**Comment by matanwerbner** [comment]
Before:
screenshot 1

After:

screenshot 2

File: app/helpers/dashboard_helper.rb [hunk 1]
```diff
@@ -49,6 +49,17 @@ def count_reports(hosts)
     data
   end
 
+  def get_overview_json(report, options = {})
+    data = [
+      [_('Active'), report[:active_hosts_ok_enabled],report_color[:active_hosts_ok_enabled]],
```
Comment (ohadlevy) [inline app/helpers/dashboard_helper.rb]: nitpick, use spaces after ',' - not sure why hound is not complaining :)

File: app/helpers/dashboard_helper.rb [hunk 2]
```diff
@@ -49,6 +49,17 @@ def count_reports(hosts)
...
+      [_('Error'), report[:bad_hosts_enabled], report_color[:bad_hosts_enabled]],
+      [_('OK'), report[:ok_hosts_enabled], report_color[:ok_hosts_enabled]],
+      [_('Pending changes'), report[:pending_hosts_enabled],report_color[:pending_hosts_enabled]],
+      [_('Out of sync'), report[:out_of_sync_hosts_enabled],report_color[:out_of_sync_hosts_enabled]],
+      [_('No report'), report[:reports_missing],report_color[:reports_missing]],
+      [_('Notification disabled'), report[:disabled_hosts],report_color[:disabled_hosts]]
+    ].to_json
+  end
   def render_overview(report, options = {})
```
Comment (tbrisker) [inline app/helpers/dashboard_helper.rb]: This function can be removed, it is no longer used afaict.

File: app/helpers/dashboard_helper.rb [hunk 3]
```diff
@@ -49,6 +49,17 @@ def count_reports(hosts)
     data
   end
 
+  def get_overview_json(report, options = {})
+    data = [
```
Comment (tbrisker) [inline app/helpers/dashboard_helper.rb]: Why did you change the array elements from hashes to arrays? it isn't clear what each position means, and you process it again in the js code, if you need the data in a certain format it makes sense to just send it in the way you need it rather then process it again later.

File: app/helpers/dashboard_helper.rb [hunk 4]
```diff
@@ -49,6 +49,17 @@ def count_reports(hosts)
...
+      [_('Active'), report[:active_hosts_ok_enabled],report_color[:active_hosts_ok_enabled]],
+      [_('Error'), report[:bad_hosts_enabled], report_color[:bad_hosts_enabled]],
+      [_('OK'), report[:ok_hosts_enabled], report_color[:ok_hosts_enabled]],
+      [_('Pending changes'), report[:pending_hosts_enabled],report_color[:pending_hosts_enabled]],
```
Comment (tbrisker) [inline app/helpers/dashboard_helper.rb]: nitpick: missing space after comma here and below

File: webpack/assets/javascripts/services/statisticsChartService.js [hunk 5]
```diff
@@ -21,6 +22,37 @@ function getChartConfig(details) {
   return config;
 }
 
+function getChartConfigByData(_data, id, type) {
```
Comment (tbrisker) [inline webpack/assets/javascripts/services/statisticsChartService.js]: If i understand correctly, this looks like a lot of complicated work to parse the data into the C3 format instead of sending it in the right format in the first place?

File: webpack/assets/javascripts/react_app/components/charts/PieChart/index.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+import ChartService from '../../../../services/statisticsChartService';
+import Chart from '../Chart';
+import uuidV1 from 'uuid/v1';
+require('./PieChart.scss');
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/charts/PieChart/index.js]: nitpick: missing a blank line between the imports and the function

**Review by tbrisker — CHANGES_REQUESTED** [review]

File: app/helpers/dashboard_helper.rb (hunk 3 above)
Comment (matanwerbner) [reply app/helpers/dashboard_helper.rb]: I think the server should be agnostic to the format of the UI charting library, it should just return the data to be shown in a reasonable way.
that said, c3 accepts data in arrays, see example:
http://c3js.org/samples/chart_pie.html
so passing it from the server that way makes sense.

File: webpack/assets/javascripts/services/statisticsChartService.js (hunk 5 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/services/statisticsChartService.js]: the server should not know about "right format", as the desired format can change if, say, we change from c3 to other charting library. 
it should just send the data in a reasonable way.

**Comment by matanwerbner** [comment]
anything else? @ohadlevy ?

File: app/helpers/dashboard_helper.rb [hunk 7]
```diff
@@ -49,15 +49,16 @@ def count_reports(hosts)
...
-            {:label=>_('Error'), :data =>report[:bad_hosts_enabled], :color => report_color[:bad_hosts_enabled]},
-            {:label=>_('OK'), :data =>report[:ok_hosts_enabled],:color => report_color[:ok_hosts_enabled]},
-            {:label=>_('Pending changes'), :data =>report[:pending_hosts_enabled],:color => report_color[:pending_hosts_enabled]},
-            {:label=>_('Out of sync'), :data =>report[:out_of_sync_hosts_enabled],:color => report_color[:out_of_sync_hosts_enabled]},
-            {:label=>_('No report'), :data =>report[:reports_missing],:color => report_color[:reports_missing]},
-            {:label=>_('Notification disabled'), :data =>report[:disabled_hosts],:color => report_color[:disabled_hosts]}]
-    flot_pie_chart 'overview', options[:custom_title], data, options.merge(:search => "search_by_legend")
+  def get_overview_json(report, options = {})
+    [
+      [_('Active'), report[:active_hosts_ok_enabled],report_color[:active_hosts_ok_enabled]],
```
Comment (ohadlevy) [inline app/helpers/dashboard_helper.rb]: nitpick, missing whitespaces after comma. e.g.
```ruby
[_('Active'), report[:active_hosts_ok_enabled], report_color[:active_hosts_ok_enabled]],
```

File: webpack/assets/javascripts/react_app/common/MountingService.js [hunk 8]
```diff
@@ -3,11 +3,16 @@ import StatisticsChartsList from '../components/charts/StatisticsChartsList';
 import PowerStatus from '../components/hosts/powerStatus/';
 import NotificationContainer from '../components/notifications/';
 import ToastsList from '../components/toastNotifications/';
+import PieChart from '../components/charts/PieChart/';
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/MountingService.js]: can you add it as a story too?

File: webpack/assets/javascripts/services/statisticsChartService.js [hunk 9]
```diff
@@ -21,6 +22,37 @@ function getChartConfig(details) {
...
+  const filteredData = _data.filter(d => d[1] !== 0);
+  const data = filteredData.reduce(
+    (curr, next) => {
+      const key = next[0];
+      const names = {
+        ...curr.names,
+        [key]: key.replace(/_/g, ' ')
```
Comment (ohadlevy) [inline webpack/assets/javascripts/services/statisticsChartService.js]: why is this needed? i assume the server can provide the correct answer without _ ?

File: webpack/assets/javascripts/services/statisticsChartService.js [hunk 10]
```diff
@@ -21,6 +22,37 @@ function getChartConfig(details) {
...
+      };
+
+      const colors = {
+        ...curr.colors,
+        [key]: next[2]
```
Comment (ohadlevy) [inline webpack/assets/javascripts/services/statisticsChartService.js]: whats next[2]? can you add a test so its a bit more understood? overall I'm not sure i understand what this code is doing and why its needed, thanks!

**Review by ohadlevy — CHANGES_REQUESTED** [review]
@matanwerbner overall looking good, a few inline commments - thanks.

File: webpack/assets/javascripts/services/statisticsChartService.js (hunk 10 above)
Comment (matanwerbner) [reply webpack/assets/javascripts/services/statisticsChartService.js]: in the server side, data is sent in the format [ [ key, value, color], [ key, value, color] ] . 
c3  accepts data in a uniqe way. colors and column names are passed by a dictionary, so reduce turns the array from the server into this specific format.
i've added a bit of comments to make things more readable hopefully...

**Comment by matanwerbner** [comment]
This PR now includes some refactoring of the old charts code to seperate into reusable components (chart, pieChart), and specific composed components (chartBox, chartModal, statisticsChartList) that were used in the statistics page.

**Comment by ohadlevy** [comment]
thanks @matanwerbner
